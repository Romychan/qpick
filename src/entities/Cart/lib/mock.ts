// eslint-disable-next-line @conarti/feature-sliced/layers-slices
import { Product } from '~/entities/Product/@x';

import { CartItem } from '../model/types';

/** Mock product for the cart */
export const MOCK_CART_PRODUCT: Product = {
  id: 1,
  title: 'Product',
  rate: 4.2,
  image: '/images/catalog/headphone-1.png',
  price: 1200,
  oldPrice: 1300,
  description: 'Test',
  category: 'Наушники',
};

/** List of mock cart items */
export const MOCK_CART: CartItem[] = [
  {
    product: MOCK_CART_PRODUCT,
    quantity: 2,
  },
];
