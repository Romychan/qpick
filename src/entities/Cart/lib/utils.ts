// eslint-disable-next-line @conarti/feature-sliced/layers-slices
import { Product } from '~/entities/Product/@x';

import { CartItem } from '../model/types';

/**
 * Add a product to the cart
 *
 * @param cartItems The current state of the cart
 * @param product Product to add to cart
 * @returns Updated state of the product cart
 */
export const addToCart = (cartItems: CartItem[], product: Product) => {
  const productInCart = cartItems.find(
    (item) => item.product.id === product.id,
  );

  if (!productInCart) {
    return [...cartItems, { product, quantity: 1 }];
  } else {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity + 1 }
        : item,
    );
  }
};

/**
 * Delete a product from the cart
 *
 * @param cartItems The current state of the cart
 * @param id The ID of the product to be deleted from the cart
 * @returns Updated state of the product cart
 */
export const deleteFromCart = (cartItems: CartItem[], id: number) => {
  return cartItems.filter((item) => item.product.id !== id);
};

/**
 * Increase the number of product in the cart
 *
 * @param cartItems The current state of the cart
 * @param id Product ID to increase the quantity
 * @returns Updated state of the product cart
 */
export const increaseProductQuantity = (cartItems: CartItem[], id: number) => {
  const productInCart = cartItems.find((item) => item.product.id === id);

  if (productInCart) {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity + 1 }
        : item,
    );
  }

  return cartItems;
};

/**
 * Decrease the number of product in the cart
 *
 * @param cartItems The current state of the cart
 * @param id Product ID to decrease the quantity
 * @returns Updated state of the product cart
 */
export const decreaseProductQuantity = (cartItems: CartItem[], id: number) => {
  const productInCart = cartItems.find((item) => item.product.id === id);

  if (productInCart) {
    return cartItems.map((item) =>
      productInCart.product.id === item.product.id
        ? { ...productInCart, quantity: item.quantity - 1 }
        : item,
    );
  }

  return cartItems;
};
