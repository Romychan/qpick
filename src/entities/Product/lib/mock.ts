import { Product } from '../model/types';

/** List of mock products */
export const MOCK_PRODUCTS: Product[] = [
  {
    id: 1,
    title: 'Apple BYZ S852I',
    rate: 4.7,
    image: '/images/catalog/headphone-1.png',
    price: 2927,
    oldPrice: 3527,
    description:
      'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
    category: 'Наушники',
  },
  {
    id: 2,
    title: 'Apple AirPods Pro 3',
    rate: 4.5,
    image: '/images/catalog/headphone-2.png',
    price: 200,
    oldPrice: null,
    description:
      "Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.",
    category: 'Наушники',
  },
  {
    id: 3,
    title: 'Apple EarPods',
    rate: 4.5,
    image: '/images/catalog/headphone-3.png',
    price: 2327,
    oldPrice: null,
    description:
      'Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.',
    category: 'Наушники',
  },
  {
    id: 4,
    title: 'Apple BYZ S123I',
    rate: 4.7,
    image: '/images/catalog/headphone-1.png',
    price: 1500,
    oldPrice: 1950,
    description:
      'Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur.',
    category: 'Наушники',
  },
  {
    id: 5,
    title: 'Apple AirPods Pro 2',
    rate: 4.5,
    image: '/images/catalog/headphone-2.png',
    price: 200,
    oldPrice: null,
    description:
      'Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
    category: 'Наушники',
  },
  {
    id: 6,
    title: 'Apple EarPods 2',
    rate: 3.8,
    image: '/images/catalog/headphone-3.png',
    price: 1200,
    oldPrice: 2000,
    description:
      'Aenean maximus, nisi eget mollis dictum, leo sapien sollicitudin ipsum, quis euismod ex erat ut tellus.',
    category: 'Наушники',
  },
  {
    id: 7,
    title: 'Apple AirPods',
    rate: 4.7,
    image: '/images/catalog/headphone-4.png',
    price: 9527,
    oldPrice: null,
    description:
      'Aenean maximus, nisi eget mollis dictum, leo sapien sollicitudin ipsum, quis euismod ex erat ut tellus.',
    category: 'Беспроводные наушники',
  },
  {
    id: 8,
    title: 'GERLAX GH-04',
    rate: 4.7,
    image: '/images/catalog/headphone-5.png',
    price: 6527,
    oldPrice: null,
    description:
      'Sed quis sem vitae justo feugiat tincidunt. Maecenas ut leo nisl. Praesent finibus ultricies ex et tristique.',
    category: 'Беспроводные наушники',
  },
  {
    id: 9,
    title: 'BOROFONE BO4',
    rate: 4.7,
    image: '/images/catalog/headphone-6.png',
    price: 7527,
    oldPrice: null,
    description:
      "Lorem Ipsum has been the industry's standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book.",
    category: 'Беспроводные наушники',
  },
];
