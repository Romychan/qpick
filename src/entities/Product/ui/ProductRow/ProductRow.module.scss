@import '~/app/assets/styles/_typography.scss';

.container {
  position: relative;
  display: flex;
  flex-direction: column;
  gap: 1rem;
}

.content {
  width: 100%;
  display: flex;
  align-items: center;
  gap: 1.6rem;
}

.image {
  aspect-ratio: 1/1;
  display: block;
  width: 10rem;
  min-width: 80px;

  & img {
    height: 100%;
    width: 100%;
    object-fit: contain;
    object-position: center;
    transition: 0.2s;
  }
}

.information {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  gap: 0.8rem;
}

.name {
  @include text-md;
  @include text-medium;
  color: var(--gray-800);
}

.price {
  @include text-sm;
  @include text-medium;
  color: var(--gray-400);
}

.row {
  display: flex;
  justify-content: space-between;
}

.actions {
  position: absolute;
  top: 0.8rem;
  right: 0.8rem;
}

@media screen and (max-width: 380px) {
  .content {
    flex-direction: column;
  }

  .image {
    width: 7rem;
    align-self: center;
  }

  .information {
    width: 100%;
    flex-direction: row;
    align-items: center;
    justify-content: space-between;
    margin-right: 0rem;
    margin-bottom: 0.5rem;
  }

  .name {
    margin-right: 0.5rem;
  }

  .row {
    width: 100%;
    display: flex;
    flex-direction: row-reverse;
    align-items: center;
    justify-content: space-between;
  }

  .price {
    align-items: flex-start;
  }
}
