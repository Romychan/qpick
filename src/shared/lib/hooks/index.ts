export { useEventListener } from './useEventListener';
export { useIsMounted } from './useIsMounted';
