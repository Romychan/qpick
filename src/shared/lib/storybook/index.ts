export { withReactHookFormDecorator } from './decorators/withReactHookFormDecorator';
export { withHeightDecorator } from './decorators/withHeightDecorator';
export { withGapDecorator } from './decorators/withGapDecorator';
export { withWidthDecorator } from './decorators/withWidthDecorator';
export { withRouterDecorator } from './decorators/withRouterDecorator';
