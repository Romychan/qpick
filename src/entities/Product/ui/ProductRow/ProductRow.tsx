import { ReactNode } from 'react';

import { formatPrice } from '~/shared/lib/utils';
import { Card } from '~/shared/ui/Card';

import { Product } from '../../model/types';

import styles from './ProductRow.module.scss';

interface IProductRowProps extends Omit<Product, 'rate'> {
  /** The bottom slot of the component for interactive elements */
  bottomSlot?: ReactNode;
  /** The slot of the component for interactive elements */
  actions?: ReactNode;
}

/** This component is used to get brief information about the product in the form of a row. */
export const ProductRow = ({
  title,
  image,
  price,
  bottomSlot,
  actions,
}: IProductRowProps) => {
  return (
    <Card className={styles.container} data-testid="product-row">
      <div className={styles.content}>
        <div className={styles.image}>
          <img src={image} alt={title} width={150} height={150} />
        </div>

        <div className={styles.information}>
          <p className={styles.name}>{title}</p>
          <p className={styles.price}>{formatPrice(price)}</p>
        </div>
      </div>

      <div className={styles.row}>{bottomSlot}</div>

      <div className={styles.actions}>{actions}</div>
    </Card>
  );
};
