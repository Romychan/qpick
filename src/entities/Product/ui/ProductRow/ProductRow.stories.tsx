import type { Meta, StoryObj } from '@storybook/react';

import {
  withRouterDecorator,
  withWidthDecorator,
} from '~/shared/lib/storybook';
import { Button, Default as DefaultButton } from '~/shared/ui/Button';

import { ProductRow } from './ProductRow';

const meta = {
  title: 'Entities/Product/ProductRow',
  component: ProductRow,
  parameters: {
    layout: 'centered',
    width: '540px',
  },
  args: {
    id: 1,
    title: 'Product Name',
    image: './images/catalog/headphone-1.png',
    price: 1200,
  },
  decorators: [withRouterDecorator, withWidthDecorator],
  tags: ['autodocs'],
} satisfies Meta<typeof ProductRow>;

export default meta;
type Story = StoryObj<typeof meta>;

/** This state is used to display the default. */
export const Default: Story = {
  args: {
    bottomSlot: 'Bottom Slot',
    actions: <Button {...DefaultButton.args}>Actions</Button>,
  },
};

/** Component without bottom and actions slot */
export const EmptySlots: Story = {
  args: {},
};
