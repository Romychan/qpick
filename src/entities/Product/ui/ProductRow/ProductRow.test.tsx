import { RenderResult, screen } from '@testing-library/react';

import { renderWithRouter } from '~/shared/lib/tests';

import { ProductRow } from './ProductRow';

describe('ProductRow', () => {
  let container: RenderResult;

  beforeEach(() => {
    container = renderWithRouter(
      <ProductRow
        id={1}
        title="Test"
        image="./test.png"
        price={1200}
        bottomSlot={<button>Bottom slot</button>}
        actions={<button>Actions</button>}
      />,
    );
  });

  it('should render correctly', () => {
    const centerRowSlotElement = screen.getByRole('button', {
      name: /Bottom/i,
    });
    const actionsElement = screen.getByRole('button', { name: /Actions/i });

    expect(centerRowSlotElement).toBeInTheDocument();
    expect(actionsElement).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    expect(container).toMatchSnapshot();
  });
});
