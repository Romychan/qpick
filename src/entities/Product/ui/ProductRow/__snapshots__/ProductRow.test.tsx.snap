// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductRow > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_1d0052 _container_f0394b"
        data-testid="product-row"
      >
        <div
          class="_content_f0394b"
        >
          <div
            class="_image_f0394b"
          >
            <img
              alt="Test"
              height="150"
              src="./test.png"
              width="150"
            />
          </div>
          <div
            class="_information_f0394b"
          >
            <p
              class="_name_f0394b"
            >
              Test
            </p>
            <p
              class="_price_f0394b"
            >
              1 200 ₽
            </p>
          </div>
        </div>
        <div
          class="_row_f0394b"
        >
          <button>
            Bottom slot
          </button>
        </div>
        <div
          class="_actions_f0394b"
        >
          <button>
            Actions
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_1d0052 _container_f0394b"
      data-testid="product-row"
    >
      <div
        class="_content_f0394b"
      >
        <div
          class="_image_f0394b"
        >
          <img
            alt="Test"
            height="150"
            src="./test.png"
            width="150"
          />
        </div>
        <div
          class="_information_f0394b"
        >
          <p
            class="_name_f0394b"
          >
            Test
          </p>
          <p
            class="_price_f0394b"
          >
            1 200 ₽
          </p>
        </div>
      </div>
      <div
        class="_row_f0394b"
      >
        <button>
          Bottom slot
        </button>
      </div>
      <div
        class="_actions_f0394b"
      >
        <button>
          Actions
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
